src/api/lib/authenticator.rb

```diff
@@ -25,7 +29,8 @@ class AdminUserRequiredError < APIException
 
   attr_reader :request, :session, :user_permissions, :http_user
 
-  def initialize(request, session)
+  def initialize(request, session, controller)
```

Comment:
Instead of injecting the entire controller here can you just have `response` as the dependency?